src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
-                      napi_destruct destructor, napi_weakref* handle) {
+napi_status napi_wrap(napi_env e,
+                      napi_value jsObject,
+                      void* nativeObj,
+                      napi_finalize finalize_cb,
+                      napi_ref* result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(jsObject);
+  CHECK_ARG(result);
 
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj;
-  CHECK_TO_OBJECT(context, obj, jsObject);
 
-  v8impl::ObjectWrapWrapper* wrap =
-      new v8impl::ObjectWrapWrapper(obj, nativeObj, destructor);
+  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+  assert(obj->InternalFieldCount() > 0);
+  obj->SetAlignedPointerInInternalField(0, nativeObj);
 
-  if (handle)
-  {
-    return napi_create_weakref(
-      e,
-      v8impl::JsValueFromV8LocalValue(wrap->handle()),
-      handle);
-  }
+  v8impl::Reference* reference = new v8impl::Reference(
```

Comment:
> Perhaps this function should never return the reference, and just keep it internal and self-deleting only for the finalizer. And if the caller needs a reference, they can create one separately.

That seems good to me. Create an internal reference if there's a finalizer, otherwise make the caller explicitly request their own weak handles. Perhaps it was initially done via that optional weak handle to avoid a second function call? Or avoid having to build a second reference, though they don't seem particularly heavy-weight...